### PR TITLE
Add --gn and --ninja to npm run build

### PR DESCRIPTION
**Pull request by spinda**
Use `--gn` to pass through ad-hoc gn args. `--gn=is_component_build:true` translates to `is_component_build=true`.

Use `--ninja` to add command line options for Ninja. `--ninja=j:48` translates to `-j 48`.

Fix https://github.com/brave/brave-browser/issues/5184

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions.

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by simonhong** [comment]
I'll review after my holiday(7/1~7/2)  :)

**Comment by simonhong** [comment]
@spinda How about doing like this - https://github.com/brave/brave-browser/pull/2559 instead of adding individually? cc @bridiver

**Comment by spinda** [comment]
@simonhong That makes a lot of sense, yeah. How does this sound:

- Unrecognized arguments in the form `--gn:is_component_build=true` get translated to gn args in the form `is_component_build=true`.
- Unrecognized arguments in the form `--ninja:j=48` get translated to Ninja command line options in the form `-j 48`.

**Comment by spinda** [comment]
Alternatively, instead of handling unrecognized argument handling to `npm run build`, I could add two new arguments: `--gn` and `--ninja`, written line `--gn=is_component_build:true` and `--ninja=j:48`

**Comment by simonhong** [comment]
> Alternatively, instead of adding unrecognized argument handling to `npm run build`, I could add two new arguments: `--gn` and `--ninja`, written like `--gn=is_component_build:true` and `--ninja=j:48`

Looks nice to me 👍 I think these options would be very helpful for building with temporal gn/ninja configurations.

**Comment by spinda** [comment]
@simonhong Okay, I've implemented that now.

File: lib/config.js [hunk 1]
```diff
@@ -16,6 +16,19 @@ const getNPMConfig = (path) => {
     process.env[package_prefix + key]
 }
 
+const parseExtraInputs = (inputs, accumulator, callback) => {
+  for (let input of inputs) {
+    let separatorIndex = input.indexOf(':')
```
Comment (bridiver) [inline lib/config.js]: `split`?

**Comment by bridiver** [comment]
@spinda in the future please add all recommended reviewers to the list

File: lib/config.js (hunk 1 above)
Comment (spinda) [reply lib/config.js]: There could be colons in the value as well. We only want to extract up to the first colon.

**Comment by spinda** [comment]
@bridiver Can I merge this?

**Pull request merged**